api/handler/get.go

```diff
@@ -180,46 +192,52 @@ func (h *handler) GetObjectHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	t := &layer.ObjectVersion{
-		BktInfo:    bktInfo,
-		ObjectName: info.Name,
-	}
-
-	if bktSettings.VersioningEnabled() {
-		t.VersionID = info.VersionID()
-	}
-
-	tagSet, lockInfo, err := h.obj.GetObjectTaggingAndLock(r.Context(), t)
-	if err != nil && !s3errors.IsS3Error(err, s3errors.ErrNoSuchKey) {
-		h.logAndSendError(w, "could not get object meta data", reqInfo, err)
-		return
-	}
-
 	if layer.IsAuthenticatedRequest(r.Context()) {
 		overrideResponseHeaders(w.Header(), reqInfo.URL.Query())
 	}
 
-	if err = h.setLockingHeaders(bktInfo, lockInfo, w.Header()); err != nil {
+	if err = h.setLockingHeaders(bktInfo, comprehensiveObjectInfo.LockInfo, w.Header()); err != nil {
 		h.logAndSendError(w, "could not get locking info", reqInfo, err)
 		return
 	}
 
-	writeHeaders(w.Header(), r.Header, extendedInfo, len(tagSet), bktSettings.Unversioned())
+	writeHeaders(w.Header(), r.Header, info, len(comprehensiveObjectInfo.TagSet), bktSettings.Unversioned())
 	if params != nil {
 		writeRangeHeaders(w, params, info.Size)
 	} else {
 		w.WriteHeader(http.StatusOK)
 	}
 
-	getParams := &layer.GetObjectParams{
-		ObjectInfo: info,
-		Writer:     w,
-		Range:      params,
-		BucketInfo: bktInfo,
-		Encryption: encryptionParams,
+	if params != nil || encryptionParams.Enabled() {
+		// unfortunately this reader is useless for us in this case, we have to re-read another one.
```

Comment:
It can be reused for sure (just piping into the decryption layer), but that's not our current primary target.